webapp/tests/Utils/Scoreboard/ScoreboardIntegrationTest.php

```diff
@@ -0,0 +1,281 @@
+     */
+    private $dj;
+
+    /**
+     * @var \App\Service\ScoreboardService
+     */
+    private $ss;
+
+    /**
+     * @var \Doctrine\ORM\EntityManager
+     */
+    private $em;
+
+    /**
+     * @var App\Entity\Contest
+     */
+    private $contest;
+
+    /**
+     * @var App\Entity\Judgehost
+     */
+    private $judgehost;
+
+    /**
+     * @var App\Entity\Problem[]
+     */
+    private $problems;
+
+    /**
+     * @var App\Entity\Team[]
+     */
+    private $teams;
+
+    protected function setUp()
+    {
+        self::bootKernel();
+
+        $this->dj = self::$container->get(DOMJudgeService::class);
+        $this->ss = self::$container->get(ScoreboardService::class);
+        $this->em = self::$container->get('doctrine')->getManager();
+
+        // Create a contest, problems and teams for which to test the
+        // scoreboard. These get deleted again in tearDown().
+        $this->contest = new Contest();
+        $this->contest
+            ->setExternalid(self::CONTEST_NAME)
+            ->setName(self::CONTEST_NAME)
+            ->setShortname(self::CONTEST_NAME)
+            ->setStarttimeString(date('Y').'-01-01 10:00:00 Europe/Amsterdam')
+            ->setActivatetimeString('-01:00')
+            ->setEndtimeString('+02:00');
+        $this->em->persist($this->contest);
+
+        $hostname = self::CONTEST_NAME.'-judgehost';
+        $this->judgehost = $this->em->getRepository(Judgehost::class)
+            ->findOneBy(['hostname' => $hostname]);
+        if ( !$this->judgehost ) {
+            $this->judgehost = new Judgehost();
+            $this->judgehost
+                ->setHostname($hostname);
+            $this->em->persist($this->judgehost);
+        }
+
+        $category = $this->em->getRepository(TeamCategory::class)
+            ->findOneBy(['sortorder' => 0]);
+
+        for($i=0; $i<self::NUM_TEAMS; $i++) {
+            $this->teams[$i] = new Team();
+            $this->teams[$i]
+                ->setName('Team '.$i)
+                ->setCategory($category);
+            $this->em->persist($this->teams[$i]);
+       }
+
+        for($i=0; $i<self::NUM_PROBLEMS; $i++) {
+            $letter = chr(ord('a') + $i);
+            $this->problems[$i] = new Problem();
+            $this->problems[$i]
+                ->setName('Problem '.$letter)
+                ->setTimelimit(5);
+
+            $cp = new ContestProblem();
+            $cp
+                ->setShortname($letter)
+                ->setProblem($this->problems[$i])
+                ->setContest($this->contest);
+
+            $this->contest->addProblem($cp);
+            $this->problems[$i]->addContestProblem($cp);
+
+            $this->em->persist($this->problems[$i]);
+            $this->em->persist($cp);
+        }
+
+        $this->em->flush();
+    }
+
+    protected function tearDown()
+    {
+        // Preserve the data for inspection if a test failed.
+        if ( !$this->hasFailed() ) {
```

Comment:
how useful is this? This test suite has multiple methods and if one fails it will still continue running through the other methods, right?

Reply:
No, since the data is then not deleted, the next test `setUp()` will fail because of the contest name that already exists and bail out.